File: skills/asset-validation/scripts/envprep.py
import json
import os
import shutil
import tempfile
import time
from pathlib import Path
from typing import Optional

from . import db


def _tmp_root() -> Path:
    for key in ("ACCEPTANCE_TMPDIR", "TMPDIR", "TMP", "TEMP"):
        value = os.environ.get(key)
        if value:
            return Path(value)
    return Path(tempfile.gettempdir())


def make_sandbox(round_tag) -> Path:
    base = _tmp_root() / f"acc-{round_tag}"
    base.mkdir(parents=True, exist_ok=True)
    return base


def rsync_fixture(fixture_path, sandbox) -> Optional[Path]:
    if not fixture_path:
        return None
    src = Path(fixture_path)
    if not src.exists():
        return None
    dest = Path(sandbox) / src.name
    shutil.copytree(src, dest, dirs_exist_ok=True)
    return dest


def _ensure_json_object(path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists():
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            if isinstance(data, dict):
                return
        except (OSError, json.JSONDecodeError):
            pass
    path.write_text("{}\n", encoding="utf-8")


def _sandbox_claude_settings(home: Path) -> dict:
    """Copy only settings needed for auth/model/permissions into sandbox."""
    src = home / ".claude" / "settings.json"
    try:
        data = json.loads(src.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        data = {}
    if not isinstance(data, dict):
        data = {}
    return {
        key: data[key]
        for key in ("env", "permissions", "model")
        if key in data
    }


def prepare_round_environment(sandbox, source_home: Optional[Path] = None) -> dict:
    sb = Path(sandbox)
    acc_home = sb / ".aut-acceptance"
    iso = sb / ".iso"
    tmp = sb / ".tmp"
    home = Path(source_home) if source_home else Path.home()
    settings = iso / "claude-settings.json"
    try:
        depth = int(os.environ.get("ACCEPTANCE_DEPTH", "0")) + 1
    except ValueError:
        depth = 1
    for path in (acc_home, iso, tmp):
        path.mkdir(parents=True, exist_ok=True)
    _ensure_json_object(settings)
    settings.write_text(
        json.dumps(_sandbox_claude_settings(home), ensure_ascii=False, indent=2)
        + "\n",
        encoding="utf-8",
    )
    return {
        "ACCEPTANCE_SANDBOX": str(sb),
        "ACCEPTANCE_HOME": str(acc_home),
        "ACCEPTANCE_TMPDIR": str(tmp),
        "ACCEPTANCE_DEPTH": str(depth),
        "TMPDIR": str(tmp),
        "TMP": str(tmp),
        "TEMP": str(tmp),
        "HOME": str(home),
        "CMDAI_CODEX_MARKETPLACE_ROOT": str(iso / "codex-marketplace"),
        "CMDAI_CLAUDE_MARKETPLACE_ROOT": str(iso / "claude-marketplace"),
        "CMDAI_CODEX_AGENTS_ROOT": str(iso / "codex-agents"),
        "CMDAI_CLAUDE_SETTINGS_PATH": str(settings),
        "BH_PROFILE_ROOT": str(iso / "bh-profiles"),
    }


def isolation_env(sandbox) -> dict:
    return prepare_round_environment(sandbox)


def cleanup_sandbox(sandbox) -> dict:
    sb = Path(sandbox)
    tmp = sb / ".tmp"
    nested = _cleanup_nested_sandboxes(sb)
    existed = sb.exists()
    tmp_existed = tmp.exists()
    if existed:
        _rmtree_retry(sb)
    return {
        "removed": str(sb),
        "existed": existed,
        "tmpdir": str(tmp),
        "tmpdir_existed": tmp_existed,
        "tmpdir_removed": tmp_existed and not tmp.exists(),
        "nested_sandboxes": nested,
    }


def _cleanup_nested_sandboxes(sandbox: Path) -> list:
    state_db = sandbox / ".aut-acceptance" / "state.sqlite3"
    return [
        _cleanup_nested_sandbox(path, sandbox)
        for path in db.round_sandbox_paths_from(state_db)
        if _is_safe_nested_sandbox(path, sandbox)
    ]


def _cleanup_nested_sandbox(path, parent_sandbox: Path) -> dict:
    nested = Path(path)
    existed = nested.exists()
    if existed:
        _rmtree_retry(nested)
    return {
        "path": str(nested),
        "existed": existed,
        "removed": existed and not nested.exists(),
    }


def _is_safe_nested_sandbox(path, parent_sandbox: Path) -> bool:
    nested = Path(path)
    if not nested.is_absolute() or not nested.name.startswith("acc-"):
        return False
    try:
        nested_resolved = nested.resolve()
        parent_resolved = parent_sandbox.resolve()
        tmp_resolved = _tmp_root().resolve()
    except OSError:
        return False
    if nested_resolved == parent_resolved:
        return False
    return (
        parent_resolved in nested_resolved.parents
        or tmp_resolved in nested_resolved.parents
        or nested_resolved.parent == tmp_resolved
    )


def _rmtree_retry(path: Path, attempts: int = 4) -> None:
    for attempt in range(attempts):
        try:
            shutil.rmtree(path)
            return
        except FileNotFoundError:
            return
        except OSError:
            if attempt + 1 >= attempts:
                raise
            time.sleep(0.1 * (attempt + 1))

File: skills/asset-validation/scripts/envprep.ts
import {
  cpSync,
  existsSync,
  mkdirSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { homedir, tmpdir } from "node:os";
import { basename, isAbsolute, join, resolve } from "node:path";

import * as db from "./db.ts";
import { parsePythonInteger } from "./number-utils.ts";
import { isStrictDescendant, stablePath } from "./path-utils.ts";
import { sleepSeconds } from "./process.ts";

export interface IsolationEnv {
  [key: string]: string;
  ACCEPTANCE_SANDBOX: string;
  ACCEPTANCE_HOME: string;
  ACCEPTANCE_TMPDIR: string;
  ACCEPTANCE_DEPTH: string;
  TMPDIR: string;
  TMP: string;
  TEMP: string;
  HOME: string;
  CMDAI_CODEX_MARKETPLACE_ROOT: string;
  CMDAI_CLAUDE_MARKETPLACE_ROOT: string;
  CMDAI_CODEX_AGENTS_ROOT: string;
  CMDAI_CLAUDE_SETTINGS_PATH: string;
  BH_PROFILE_ROOT: string;
}

export function temporaryRoot(): string {
  for (const key of ["ACCEPTANCE_TMPDIR", "TMPDIR", "TMP", "TEMP"] as const) {
    const value = process.env[key];
    if (value) return value;
  }
  return tmpdir();
}

export function makeSandbox(roundTag: string): string {
  const base = join(temporaryRoot(), `acc-${roundTag}`);
  mkdirSync(base, { recursive: true });
  return base;
}

export function rsyncFixture(fixturePath: string | null, sandbox: string): string | null {
  if (!fixturePath || !existsSync(fixturePath)) {
    return null;
  }
  const destination = join(sandbox, basename(fixturePath));
  cpSync(fixturePath, destination, {
    recursive: true,
    force: true,
    dereference: true,
  });
  return destination;
}

function ensureJsonObject(path: string): void {
  mkdirSync(join(path, ".."), { recursive: true });
  if (existsSync(path)) {
    try {
      const data = JSON.parse(readFileSync(path, "utf8"));
      if (typeof data === "object" && data !== null && !Array.isArray(data)) {
        return;
      }
    } catch {
      // Replace invalid or unreadable settings with an empty JSON object.
    }
  }
  writeFileSync(path, "{}\n", "utf8");
}

function sandboxClaudeSettings(home: string): Record<string, unknown> {
  let data: unknown = {};
  try {
    data = JSON.parse(readFileSync(join(home, ".claude", "settings.json"), "utf8"));
  } catch {
    data = {};
  }
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    data = {};
  }
  const source = data as Record<string, unknown>;
  const output: Record<string, unknown> = {};
  for (const key of ["env", "permissions", "model"]) {
    if (key in source) output[key] = source[key];
  }
  return output;
}

function parseDepth(): number {
  const raw = process.env.ACCEPTANCE_DEPTH ?? "0";
  const parsed = parsePythonInteger(raw);
  return parsed === null ? 1 : parsed + 1;
}

export function prepareRoundEnvironment(
  sandbox: string,
  sourceHome?: string,
): IsolationEnv {
  const acceptanceHome = join(sandbox, ".aut-acceptance");
  const isolation = join(sandbox, ".iso");
  const temporary = join(sandbox, ".tmp");
  const home = sourceHome || homedir();
  const settings = join(isolation, "claude-settings.json");
  for (const path of [acceptanceHome, isolation, temporary]) {
    mkdirSync(path, { recursive: true });
  }
  ensureJsonObject(settings);
  writeFileSync(
    settings,
    `${JSON.stringify(sandboxClaudeSettings(home), null, 2)}\n`,
    "utf8",
  );
  const herdrEnvironment = Object.fromEntries(
    ["HERDR_ENV", "HERDR_PANE_ID", "HERDR_TAB_ID", "HERDR_WORKSPACE_ID", "HERDR_SOCKET_PATH"]
      .flatMap((key) => process.env[key] ? [[key, process.env[key] as string]] : []),
  );
  return {
    ...herdrEnvironment,
    ACCEPTANCE_SANDBOX: sandbox,
    ACCEPTANCE_HOME: acceptanceHome,
    ACCEPTANCE_TMPDIR: temporary,
    ACCEPTANCE_DEPTH: String(parseDepth()),
    TMPDIR: temporary,
    TMP: temporary,
    TEMP: temporary,
    HOME: home,
    CMDAI_CODEX_MARKETPLACE_ROOT: join(isolation, "codex-marketplace"),
    CMDAI_CLAUDE_MARKETPLACE_ROOT: join(isolation, "claude-marketplace"),
    CMDAI_CODEX_AGENTS_ROOT: join(isolation, "codex-agents"),
    CMDAI_CLAUDE_SETTINGS_PATH: settings,
    BH_PROFILE_ROOT: join(isolation, "bh-profiles"),
  };
}

export function isolationEnv(sandbox: string): IsolationEnv {
  return prepareRoundEnvironment(sandbox);
}

export interface NestedSandboxCleanup {
  path: string;
  existed: boolean;
  removed: boolean;
}

export interface SandboxCleanup {
  removed: string;
  existed: boolean;
  tmpdir: string;
  tmpdir_existed: boolean;
  tmpdir_removed: boolean;
  nested_sandboxes: NestedSandboxCleanup[];
  [key: string]: unknown;
}

export function cleanupSandbox(sandbox: string): SandboxCleanup {
  const temporary = join(sandbox, ".tmp");
  const nested = cleanupNestedSandboxes(sandbox);
  const existed = existsSync(sandbox);
  const temporaryExisted = existsSync(temporary);
  if (existed) removeTreeRetry(sandbox);
  return {
    removed: sandbox,
    existed,
    tmpdir: temporary,
    tmpdir_existed: temporaryExisted,
    tmpdir_removed: temporaryExisted && !existsSync(temporary),
    nested_sandboxes: nested,
  };
}

function cleanupNestedSandboxes(sandbox: string): NestedSandboxCleanup[] {
  const stateDatabase = join(sandbox, ".aut-acceptance", "state.sqlite3");
  return db.roundSandboxPathsFrom(stateDatabase)
    .filter((path) => isSafeNestedSandbox(path, sandbox))
    .map((path) => cleanupNestedSandbox(path));
}

function cleanupNestedSandbox(path: string): NestedSandboxCleanup {
  const existed = existsSync(path);
  if (existed) removeTreeRetry(path);
  return { path, existed, removed: existed && !existsSync(path) };
}

function isSafeNestedSandbox(path: string, parentSandbox: string): boolean {
  if (!isAbsolute(path) || !basename(path).startsWith("acc-")) {
    return false;
  }
  try {
    const nested = stablePath(path) as string;
    const parent = stablePath(parentSandbox) as string;
    const temporary = stablePath(temporaryRoot()) as string;
    if (nested === parent) return false;
    return isStrictDescendant(nested, parent)
      || isStrictDescendant(nested, temporary)
      || resolve(join(nested, "..")) === resolve(temporary);
  } catch {
    return false;
  }
}

interface RemoveTreeOps {
  remove(path: string): void;
  exists(path: string): boolean;
  sleep(seconds: number): void;
}

const defaultRemoveTreeOps: RemoveTreeOps = {
  remove: (path) => rmSync(path, { recursive: true, force: true }),
  exists: existsSync,
  sleep: sleepSeconds,
};

export function removeTreeRetry(
  path: string,
  attempts = 6,
  operations: RemoveTreeOps = defaultRemoveTreeOps,
): void {
  for (let attempt = 0; attempt < attempts; attempt += 1) {
    try {
      operations.remove(path);
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code === "ENOENT") return;
      if (attempt + 1 >= attempts) throw error;
      operations.sleep(0.1 * (attempt + 1));
      continue;
    }
    // tmux kill-session can return just before a pane child exits. That child
    // may recreate an absolute sandbox path after the first successful rm.
    // Require a short absence window and retry the exact validated target.
    operations.sleep(0.1 * (attempt + 1));
    if (!operations.exists(path)) return;
  }
  throw new Error(`cleanup target reappeared after ${attempts} attempts: ${path}`);
}
